### DISCUSSION TITLE
v16网络升级清单模版探索

### DISCUSSION THREAD

**Discussion by Fatman13** [filecoin-project/venus#4848] [Ideas]
## 研发

- [x] 合并PR选取([参考](https://github.com/filecoin-project/venus/discussions/4847))
- [ ] 合并1.16.0-rcX
- [ ] Venus各组件版本号发布更新

## 测试

- [ ] 覆盖率15%？
- [ ] ButterFly网络测试成功
- [ ] Calib网络测试成功
- [ ] mainnet上线跟踪

## 部署

venus-auth, gateway, market, miner 需同时升级部署, 并有以下注意事项:
1. venus-auth中的user确保状态为激活.
2. 在gateway中, 现在增加了user和miner的关联性检查, 所以上线时, 需要做相应的检查
3. 组件升级顺序(后续添加)

### venus checklist
- [ ] ~/.venus/config.json中本次升级高度是否设置
- [ ] 对比lotus和venus接口返回结果
- [ ] 模拟主网升级，比较lotus和venus升级后root
- [ ] 能否正常同步

**Comment by Fatman13** [comment]
原文 by @zl03jsj 
### 升级背景描述

- 支持多个miner之间可共享使用token的特性
### 版本信息

| 组件             | 版本                                          |
| ---------------- | --------------------------------------------- |
| venus            | -                                 |
| venus-gateway    | -                                             |
| venus-messager   | -                                             |
| venus-auth       |                                               |
| venus-wallet     |                                               |
| venus-sealer     |                                               |
| venus-miner      |                                               |

### 组件升级顺序

```mermaid
graph LR;
auth(auth)
miner[miner]
gateway[gateway]
market[market]
auth-->gateway
gateway-->miner
gateway-->market
```

### 升级Venus-auth

- [ ] 版本信息:[tag-commit]

- 新特性
  支持多个miner共享同一个token的功能

- 数据库迁移

  之前是miner和user有1:1对应的关系, 现在变成了n:1的关系, 导致数据库结构变动.

  **迁移方式:**`部署后自动迁移`

- API变化

  - 新增API: `GetUserByMiner`,  `ListUsersWithMiners`
  - 接口变化:`ListUsers`, `GetUser`接口返回的`user`信息中不再包含对应的`miner`信息, 通过`GetUserByMiner`接口来获取

- 组件升级影响
  - [x] 破坏兼容性, `venus-market`, `venus-gateway`, `venus-miner`需同步升级

- 启动完成后检查项:
    - [ ] 日志观察
      
      - 数据库自动迁移

        - 成功日志:`migrate from:0, to:1 success.`
      
        - 失败日志:`migrate from store version:0 failed:`
      
    - [ ] 检查user状态信息, 确保被使用的`user`状态都为`enabled`状态.
      ```shell
       ./venus-auth user get 1-125
      name: 1-125
      sourceType: 0 	// miner:1
      state: enabled 	// 0: disable, 1: enable
      ```
    
    - [ ] 检查`user`和`miner`的绑定关系, 确保关联正确
      ```shell
      ./venus-auth user miner list 1-125
      user: 1-125, miner count:3
      idx  miner   create-time
      0    f01087  Tue, 22 Mar 2022 17:30:36 CST
      1    f01143  Wed, 23 Mar 2022 10:20:14 CST
      2    f01153  Tue, 12 Apr 2022 15:24:26 CST`
      ```

### 升级Venus-gateway检查

- [ ] 版本信息:[tag-commit]

- 新特性

  - 当market, sealer注册到gateway时, gateway在验证miner时, 增加了对miner的检查:

    - miner是否与其token对应的user的映射关系

    - user是否为激活状态

- 组件升级影响

  - [x] 破坏兼容性, `sealer`,`market`需同步升级

- 启动完成后检查项

  - [ ] 日志观察

    - 当`venus-sealer`注册到gateway时, 如果验证miner失败可以看到相关日志:
   ```shell
   error in RPC call to 'Gateway.ListenProofEvent': verify miner:t01087 failed:user:1-125 is disabled, please enable it on 'venus-auth'
   error in RPC call to 'Gateway.ListenMarketEvent': verify miner:t01087 failed:user:1-125 is disabled, please enable it on 'venus-auth'
   ```
  

### 升级Venus-miner
- [ ] 版本信息:[tag-commit]

- 组件升级影响
  
  - [x] 破坏兼容性, 使用新的`venus-auth` api, 需与新版本`venu-auth`同步升级
  
- 启动完成后检查项

  - [ ] 日志观察
    - venus-miner在刷新miner列表是, 新增了验证了miner对应的user是否为激活状态, 如果user未激活, 可以看到相关的日志:
    ```shell
    2022-04-11T14:49:26.636+0800    WARN    auth-miners     auth/auth.go:103        user: f01031 state is disabled, it's miners won't be updated
    2022-04-11T14:49:26.636+0800    WARN    auth-miners     auth/auth.go:103        user: 200-22 state is disabled, it's miners won't be updated
    2022-04-11T14:49:26.636+0800    WARN    auth-miners     auth/auth.go:103        user: 1.125 state is disabled, it's miners won't be updated
    2022-04-11T14:49:26.639+0800    WARN    auth-miners     auth/auth.go:103        user: zl state is disabled, it's miners won't be updated
    ```

### 升级Venus-market检查项

- [ ] 版本信息:[tag-commit]

- 组件升级影响

  - [x] 破坏兼容性, 使用新的`venus-auth` api, 需与新版本`venu-auth`同步升级

- 启动完成后检查项

  - [ ] 日志观察
    - venus-market运行在`pool`模式下, 会自动从veus-auth刷新miner列表,验证了miner对应的user是否为激活状态, 如果user未激活, 可以看到相关的日志:

  ```shell
  2022-04-11T15:17:17.929+0800    WARN    address-manager minermgr/minermgr.go:158        user:f01031 state is: disabled, won't list its mienrs
  2022-04-11T15:17:17.929+0800    WARN    address-manager minermgr/minermgr.go:158        user:200-22 state is: disabled, won't list its mienrs
  2022-04-11T15:17:17.929+0800    WARN    address-manager minermgr/minermgr.go:158        user:1.125 state is: disabled, won't list its mienrs
  2022-04-11T15:17:17.929+0800    WARN    address-manager minermgr/minermgr.go:158        user:zl state is: disabled, won't list its mienrs
  ```

### 升级方式

- [x] 集群切换式升级

  1. 线下组件迁移到阿里云上

  2. 升级华为云组件

  3. 检查华为云能否正常工作，验证华为云版本能否正常出快和推送消息

  4. 逐步迁移并升级阿里云线下节点到华为云

  5. 升级阿里云

  6. 恢复原有的部署结构

- [ ] 直接部署

**Comment by Fatman13** [comment]
整理模板:
### 升级背景描述
- 升级背景描述
### 版本信息
| 组件             | 版本          |
| ---------------- | ------------- |
| venus            | vx.x.x-commit |
| venus-gateway    | -             |
| venus-messager   | -             |
| venus-auth       |               |
| venus-wallet     |               |
| venus-sealer     |               |
| venus-miner      |               |
| chain-co     |               |
### 涉及组件升级顺序
```mermaid
graph LR;
auth(auth)
miner[miner]
gateway[gateway]
market[market]
venus(venus)
venus-->auth-->gateway
gateway-->miner
gateway-->market
```
### 升级[组件名称]
- 版本信息:[tag-commit]
- 新特性
  > [描述]

- 数据库迁移
  > 变化原因:[optional]
  > 迁移方式:[must]
  > 	[手动], 描述把工具/脚本的使用方法
  > 	[自动], 省略

- API变化
  > [发生变化的API], [新增API], [删除API]

- 启动命令参数
- 配置/配置文件
- 环境变量
- 组件升级影响
  - [ ] 是否破坏与其它组件的兼容性
    > 是: 需列出涉及的需要同步升级的组件
    > 否: 略
- 启动完成后检查项
  > 日志检查, 命令行检查等.

### 升级方式

- [x] 集群切换式升级

- [ ] 直接部署

**Reply by Fatman13** [reply]
消息上链，出块。

**Reply by Fatman13** [reply]
filscan上检查，windowPost, winningPost，

**Reply by Fatman13** [reply]
回归测试

**Comment by zl03jsj** [comment]
模板应该是用于网络版本升级时的注意检查列表.
比如:
需要先在venus-share中升级对lotus的依赖, 
生成api diff...
...
检查lotus和venus api
....
